File: src/core/SDRClassifier.py
from .SDR import *
from .SDRMap import *
from .SDRDictionary import *

class SDRClassifier:

    def __init__(self,size=8,threshold=0.5):
        self.map = SDRMap(size=size,threshold=threshold)
        self.dict = SDRDictionary()
        self.patterns = []

    def get(self,arr,match=True):
        out = self.map.get(arr)

        # if there is not match
        if(len(out) == 0):
            self.patterns.append(arr)
            sdr = SDR(2048)
            self.map.set(arr,sdr.indices)
            out = sdr.indices
            self.dict.set(out,','.join(map(str, arr)))

        # if there is more than one match
        elif(len(out) > self.map.size and match == True):
            sdrs = [[int(x) for x in i.split(',')] for i in self.dict.get(out)]
            match = SDR.Match(arr,sdrs)
            out = self.map.get(match)

        return out

File: src/core/SDRClassifier.js

const SDR = require('./SDR')
const SDRMap = require('./SDRMap')
const SDRDictionary = require('./SDRDictionary')

module.exports = class SDRClassifier {

    constructor(population=8,threshold=0.5){
        this.map = new SDRMap(population,threshold)
        this.dict = new SDRDictionary()
        this.keys = []
    }

    get(key,match=true){

        let val = this.map.get(key)
        
        // if there is more then one match
        if(val && val.length > this.map.population && match){
            const keys = this.dict.get(val).map(v => v.split(',').map(i => parseInt(i)))
            val = this.map.get(SDR.Match(key,keys))
        }

        // if there is no match
        else if(!val){
            this.keys.push(key)
            val = SDR.Random(8,2048)
            this.map.set(key,val)
            this.dict.set(val,key.join(','))
        }

        return val

    }

}